src/kademlia/sample_infohashes.cpp

```diff
@@ -0,0 +1,151 @@
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include <libtorrent/kademlia/sample_infohashes.hpp>
+#include <libtorrent/kademlia/dht_observer.hpp>
+#include <libtorrent/kademlia/node.hpp>
+#include <libtorrent/performance_counters.hpp>
+
+#include <libtorrent/aux_/numeric_cast.hpp>
+
+namespace libtorrent { namespace dht
+{
+
+sample_infohashes::sample_infohashes(
+	node& dht_node
+	, node_id const& target
+	, data_callback const& dcallback
+	, nodes_callback const& ncallback)
+	: find_data(dht_node, target, ncallback)
+	, m_data_callback(dcallback)
+{}
+
+char const* sample_infohashes::name() const { return "sample_infohashes"; }
+
+observer_ptr sample_infohashes::new_observer(udp::endpoint const& ep
+	, node_id const& id)
+{
+	auto o = m_node.m_rpc.allocate_observer<sample_infohashes_observer>(self(), ep, id);
+#if TORRENT_USE_ASSERTS
+	if (o) o->m_in_constructor = false;
+#endif
+	return o;
+}
+
+bool sample_infohashes::invoke(observer_ptr o)
+{
+	entry e;
+	e["y"] = "q";
+	entry& a = e["a"];
+
+	e["q"] = "sample_infohashes";
+	a["target"] = target().to_string();
+
+	m_node.stats_counters().inc_stats_counter(counters::dht_sample_infohashes_out);
+
+	return m_node.m_rpc.invoke(e, o->target_ep(), o);
+}
+
+void sample_infohashes::done()
+{
+#ifndef TORRENT_DISABLE_LOGGING
+	get_node().observer()->log(dht_logger::traversal, "[%u] %s DONE, response %d, timeout %d"
+		, id(), name(), num_responses(), num_timeouts());
+#endif
+
+	find_data::done();
+}
+
+void sample_infohashes::got_samples(time_duration interval
+	, int num, std::vector<sha1_hash> samples)
+{
+	if (m_data_callback) m_data_callback(interval, num, std::move(samples));
+}
+
+void sample_infohashes_observer::reply(msg const& m)
+{
+	bdecode_node r = m.message.dict_find_dict("r");
+	if (!r)
+	{
+#ifndef TORRENT_DISABLE_LOGGING
+		get_observer()->log(dht_logger::traversal, "[%p] missing response dict"
+			, static_cast<void*>(algorithm()));
+#endif
+		timeout();
+		return;
+	}
+
+	std::int64_t const interval = r.dict_find_int_value("interval", -1);
+	if (interval < 0 || interval > 21600) // TODO: put constant in a common place
+	{
+#ifndef TORRENT_DISABLE_LOGGING
+		get_observer()->log(dht_logger::traversal, "[%p] wrong or missing interval value"
+			, static_cast<void*>(algorithm()));
+#endif
+		timeout();
+		return;
+	}
+
+	std::int64_t const num = r.dict_find_int_value("num", -1);
+	if (num < 0 || num > std::numeric_limits<int>::max())
+	{
+#ifndef TORRENT_DISABLE_LOGGING
+		get_observer()->log(dht_logger::traversal, "[%p] wrong or missing num value"
+			, static_cast<void*>(algorithm()));
```

Comment:
please use ``search_id`` from the traversal_algorithm for logging instead of the pointer (I changed this a while ago to make the log output deterministic across simulation runs)